spec/lib/aliquot/payment_spec.rb

```diff
@@ -3,157 +3,130 @@
+  end
 
+  it 'rejects non-base64 shared_secret' do
     block = proc do
-      Aliquot::Payment.new(token.to_json, shared_secret, merchant_id, signing_keys: keystring)
+      Aliquot::Payment.new(generator.token.to_json,
+                       'not base64',
+                       generator.merchant_id,
+                       signing_keys: generator.extract_root_signing_keys)
+        .process
     end
+    expect(&block).to raise_error(Aliquot::InvalidSharedSecretError, 'shared_secret must be base64')
+  end
+
+  it 'rejects shared_secret when not 32 bytes' do
+    generator.shared_secret = 'not 32 bytes'
+    is_expected.to raise_error(Aliquot::InvalidSharedSecretError, 'shared_secret must be 32 bytes when base64 decoded')
+  end
 
-    expect(&block).to raise_error(Aliquot::ValidationError, /signedMessage must be valid JSON/)
+  it 'rejects when signature of signedMessage does not match' do
+    generator.signature = AliquotPay.new.build_signature
+    is_expected.to raise_error(Aliquot::InvalidSignatureError, 'signature of signedMessage does not match')
   end
 
-  it 'handles decryption error succesfully' do
+  it 'rejects when failing to verify signature' do
     name = OpenSSL::ASN1::PrintableString.new('not a signature')
     asn1 = OpenSSL::ASN1::Sequence.new([name])
     der  = asn1.to_der
 
-    @payment = AliquotPay.payment
-    token['signature'] = Base64.strict_encode64(der)
-    a = Aliquot::Payment.new(token_string, shared_secret, merchant_id, signing_keys: keystring)
-    expect { a.process }.to raise_error(Aliquot::InvalidSignatureError, /nested asn1 error/)
+    generator.signature = Base64.strict_encode64(der)
+    is_expected.to raise_error(Aliquot::InvalidSignatureError, /\Aerror verifying signature,/)
   end
-
-  # KSE: I can't figure out how to provoke an error the following cases
-  it 'fails to derive keys gracefully'
 end
 
 describe Aliquot::Payment do
-  let(:token) { AliquotPay.generate_token_ecv2(payment, key_ecv2, intermediate_key, recipient) }
-  let(:token_string) { token.to_json }
-  let(:shared_secret) { extract_shared_secret(token, recipient) }
-  let(:payment) { AliquotPay.payment }
-
-  context 'ECv2' do
-    it 'validates ECv2 token' do
-      a = Aliquot::Payment.new(token_string, shared_secret, merchant_id, signing_keys: keystring)
-      a.process
-    end
-
-    it 'requires intermediateSigningKey when ECv2' do
-      token.delete('intermediateSigningKey')
-      block = proc do
-        Aliquot::Payment.new(token_string, shared_secret, merchant_id, signing_keys: keystring)
-      end
-      expect(&block).to raise_error(Aliquot::ValidationError, /intermediateSigningKey must be filled/)
+  subject do
+    -> do Aliquot::Payment.new(generator.token.to_json,
+                               generator.shared_secret,
+                               generator.merchant_id,
+                               signing_keys: generator.extract_root_signing_keys)
+        .process
```

Comment:
This indentation differs a bit from lines 61-63 :slightly_smiling_face: I prefer this :slightly_smiling_face: